sympy/solvers/ode.py

```diff
@@ -1482,6 +1491,233 @@ def _test_term(coeff, order):
+    def _power_counting(num):
+        _pow = {0}
+        for val in num:
+            if val.has(x):
+                if isinstance(val, Pow) and val.base == x:
+                    _pow.add(val.exp)
+                elif val == x:
+                    _pow.add(1)
+                else:
+                    _pow.update(_power_counting(val.args))
+        return _pow
+
+    pow_num = _power_counting((num, ))
+    pow_dem = _power_counting((dem, ))
+    pow_dem.update(pow_num)
+
+    _pow = pow_dem
+    k = 0
+    if _pow:
+        k = list(_pow)[0]
+        # computing value of k
+        for i in _pow:
+            k = gcd(i, k)
+
+    # computing I0 of the given equation
+    I0 = powdenest(simplify(factor(((J1/k**2) - S(1)/4)/((x**k)**2))), force=True)
+    I0 = factor(simplify(powdenest(I0.subs(x, x**(S(1)/k)), force=True)))
+    num, dem = I0.as_numer_denom()
+
+    max_num_pow = max(_power_counting((num, )))
+    dem_args = dem.args
+    sing_point = []
+    dem_pow = []
+    # calculating singular point of I0.
+    for arg in dem_args:
+        if arg.has(x):
+            if isinstance(arg, Pow):
+                # (x-a)**n
+                dem_pow.append(arg.exp)
+                sing_point.append(solve(arg.base, x)[0])
+            else:
+                # (x-a) type
+                dem_pow.append(1)
+                sing_point.append(solve(arg, x)[0])
+
+    dem_pow.sort()
+    # checking if equivalance is exists or not.
+
+    if equivalance(max_num_pow, dem_pow) == "2F1":
+        return {'I0':I0, 'k':k, 'sing_point':sing_point, 'type':"2F1"}
+    else:
+        return {'I0':None, 'k':None, 'sing_point':None, 'type':None}
+
+def ode_2nd_hypergeometirc(eq, func, order, match):
+
+    from sympy.simplify.hyperexpand import hyperexpand
+    x = func.args[0]
+    f = func
+    C0, C1 = get_numbered_constants(eq, num=2)
+    a = match['a']
+    b = match['b']
+    c = match['c']
+    a3 = Wild('a3', exclude=[f, f.diff(x), f.diff(x, 2)])
+    b3 = Wild('b3', exclude=[f, f.diff(x), f.diff(x, 2)])
+    c3 = Wild('c3', exclude=[f, f.diff(x), f.diff(x, 2)])
+
+    A = cancel((match['r'][b3])/((match['r'])[a3]))
+    B = cancel((match['r'][c3])/((match['r'])[a3]))
+
+    sol = None
+    if match['type'] == "2F1":
+        if c.is_integer == False or c.is_positive is not True:
+            sol = C0*hyper([a, b], [c], x) + C1*hyper([a-c+1, b-c+1], [2-c], x)*x**(1-c)
+        elif c == 1:
+            y2 = Integral(exp(Integral((-(a+b+1)*x + c)/(x**2-x), x))/hyperexpand(hyper([a, b], [c], x)**2), x)*hyper([a, b], [c], x)
+            sol = C0*hyper([a, b], [c], x) + C1*y2
+        # applying transformation in the solution
+        subs = match['mobius']
+        dtdx = simplify(1/(subs.diff(x)))
+        _B = simplify(((a + b + 1)*x - c).subs(x, subs)*dtdx)
+        _B = simplify(_B + ((x**2 -x).subs(x, subs))*(dtdx.diff(x)*dtdx))
+        _A = simplify((x**2 - x).subs(x, subs)*(dtdx**2))
```

Comment:
I am substituting, so sometimes, we are getting a complex equation and that's why I am simplifying at each substitution as we have to find the Integral of the equation in the next step so I am trying to make it simple.

Reply:
That makes sense but we try to avoid using generic simplify in library code except in situations where we really don't know what kind of expression we have. We know in some sense the form of these equations so I think there should be a specific kind of simplification that is needed.